### PR TITLE
unbound: Fix PTR records for DHCP endpoints

### LINKED ISSUE #4597: DHCP Lease Update script does not add PTR correctly in Unbound
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues and I am convinced that mine is new.

**Describe the bug**

The [unbound_dhcp.py script](https://github.com/opnsense/core/blob/master/src/opnsense/scripts/dns/unbound_dhcpd.py) for updating DHCP leases in Unbound does not work for the PTR entries. In particular, the live update via `unbound-control` does not work for the PTR (A record works as expected). The `local-data-ptr` entry in `dhcpleases.conf` works correctly, so after a restart of Unbound, the IP is correctly resolved to the right hostname.

I believe the culprit here is the following line in the `unbound_dhcp.py` script:

https://github.com/opnsense/core/blob/bdc443096370efd38e5833eb11858d6dcdb16e5d/src/opnsense/scripts/dns/unbound_dhcpd.py#L154

The problem seems to be the format of the entry for PTR send with `unbound-control local_data`. I tried the command the script would execute while being logged into my OPNSense box with `unbound-control -c /path/here local_data "10.86.5.21 PTR amun.home"`. This returned `ok`, but `nslookup 10.86.5.21` still returned `** server can't find 21.5.86.10.in-addr.arpa: NXDOMAIN`. 

After some different ways of formatting, the following worked: `unbound-control -c /path/here local_data "21.5.86.10.in-addr.arpa IN PTR amun.home"`

**To Reproduce**

In the following reproduction, I've got a host `amun.home` which gets the DHCP IP `10.86.5.21`, dynamically.

Steps to reproduce the actual error I observed:
1. Delete lease in DHCP config
2. Wait until lease expires and is removed from Unbound
3. run `nslookup 10.86.5.21`. This fails as expected with `** server can't find 21.5.86.10.in-addr.arpa: NXDOMAIN`
4. Have the machine get a new lease
5. run `nslookup amun.home`. Observe that you get the correct IP, `10.86.5.21` back
6. run `nslookup 10.86.5.21`. This fails again with `** server can't find 21.5.86.10.in-addr.arpa: NXDOMAIN`
7. Have a look at `/var/unbound/dhcpleases.conf` and observe that the correct entry `local-data-ptr: "10.86.5.21 amun.home"` exists
8. Restart Unbound
9. run `nslookup 10.86.5.21` yet again. Observe that you now get the correct hostname

**Expected behavior**

A reverse DNS lookup on the address of a host added to Unbound via DHCP returns that host's FQDN.

**Environment**

OPNsense 20.7.7 (amd64, OpenSSL).

### PR DESCRIPTION
**Pull request by gwjo**
Changes:
* Add PTR records for DHCP endpoints in the correct format
* Delete PTR records when A records are deleted
* Use bulk add/remove requests to unbound-control to improve performance

Testing:
* Created new endpoint and confirmed hostname and reverse lookup worked
* After DHCP entry timed out and DNS entry is removed lookups for hostname and IP fail.

Fixes #4597

### CODE REVIEW

File: src/opnsense/scripts/dns/unbound_dhcpd.py [hunk 1]
```diff
@@ -119,6 +126,9 @@ def run_watcher(target_filename, domain):
                     and 'client-hostname' in lease and 'address' in lease and lease['client-hostname']:
                 cached_leases[lease['address']] = lease
                 dhcpd_changed = True
+
+        remove_rr = ''
+        add_rr = ''
```
Comment (AdSchellevis) [inline src/opnsense/scripts/dns/unbound_dhcpd.py:131]: Maybe it's better to replace both for lists, it's slightly faster and saves the \n on each line.

File: src/opnsense/scripts/dns/unbound_dhcpd.py [hunk 2]
```diff
@@ -129,8 +139,8 @@ def run_watcher(target_filename, domain):
                         syslog.LOG_NOTICE,
                         "dhcpd expired %s @ %s" % (cached_leases[address]['client-hostname'], address)
                     )
-                    unbound_control(['local_data_remove',  cached_leases[address]['client-hostname']])
                     fqdn = '%s.%s' % (cached_leases[address]['client-hostname'], domain)
+                    remove_rr += f'{fqdn}\n{ipaddress.ip_address(address).reverse_pointer}\n'
```
Comment (AdSchellevis) [inline src/opnsense/scripts/dns/unbound_dhcpd.py:143]: shouldn't it drop the A record here too?

**Review by AdSchellevis — CHANGES_REQUESTED** [review]
@gwjo thanks for working on this, I have some small remarks and questions which are in the review.

File: src/opnsense/scripts/dns/unbound_dhcpd.py (hunk 1 above)
Comment (gwjo) [reply src/opnsense/scripts/dns/unbound_dhcpd.py:131]: Sure.

Each RR record needs to be sent on a separate line, but I can do this inside of `unbound_control()`, keeping this code cleaner.

File: src/opnsense/scripts/dns/unbound_dhcpd.py (hunk 2 above)
Comment (gwjo) [reply src/opnsense/scripts/dns/unbound_dhcpd.py:143]: It does there are two lines here, the reverse pointer for the PTR record and the FQDN for the A record.

Switching to a list will make this much clearer.

**Comment by AdSchellevis** [comment]
@gwjo thanks, I seem to have missed this https://github.com/opnsense/core/commit/aa6c44e767a8a61b790d4232a478ff767d8c3293 in my last review, can you check master again?

**Comment by gwjo** [comment]
> @gwjo thanks, I seem to have missed this [aa6c44e](https://github.com/opnsense/core/commit/aa6c44e767a8a61b790d4232a478ff767d8c3293) in my last review, can you check master again?

Looks good.  

Good catch, thanks!

**Pull request merged**